src/qp_block_data.jl

```diff
@@ -0,0 +1,411 @@
+# Copyright (c) 2013: Iain Dunning, Miles Lubin, and contributors
```

Comment:
What's with the copyright here? We are unter MIT license, we can not have copyrighted stuff in here I think.

Reply:
Just talked about this at the JuMP-dev monthly call because I'm also not a copyright expert.
There is no issue with it. You could actually also add a copyright header on top of all your files and the fact that this one has a different header is not an issue

Reply:
There are two independent things here:

1. the MIT license
2. the copyright header

The two work in tandem, and are not in opposition.

Manopt.jl is MIT licensed, and so is Ipopt. There is no problem using the MIT licensed code from Ipopt in Manopt.jl. However, the Ipopt license stipulates that you must include the copyright notice in other works:
https://github.com/jump-dev/Ipopt.jl/blob/adfde744946ccb13c291e8e61440868c9cc98943/LICENSE.md?plain=1#L12-L13

A copyright header does not add anything, legally, since copyright is automatically granted. (You do not need to write "copyright (c)" for it to apply.) However, it makes things a bit clearer, and anyone copying the file is nudged to include the header.

Since Benoit modified the file, you may _also_ add an additional line to the header like:
```
# Copyright (c) 2013: Iain Dunning, Miles Lubin, and Ipopt.jl contributors
# Copyright (c) 2018-2023: Ronny Bergmann, Benoit Legat, and Manopt.jl contributors
```

You might want to read

 * https://softwareengineering.stackexchange.com/questions/317041/should-i-add-the-license-in-every-header-and-source-file
 * https://opensource.google/documentation/reference/releasing/licenses

Reply:
But both links also talk about an MIT header, that was the first thing I was confused about here; just a (c) line, which nearly reads (or for me reads) as plain copyright.

But second confusion: this file was written (?) by Benoit? Why are there then 3 other people in the header?
And this file aims to bring Manopt.jl together with MOI/JuMP. Why does the line speak about Ipopt then?

About having such a line in every code file, hm, I am not yet convinced, but is of course something easy to do (just that one has to remember to update the year every year).

Reply:
> But second confusion: this file was written (?) by Benoit? Why are there then 3 other people in the header?
And this file aims to bring Manopt.jl together with MOI/JuMP. Why does the line speak about Ipopt then?

It was written by @odow to make the Ipopt MOI interface, the other people are the initial creators of the Ipopt interface.

> About having such a line in every code file, hm, I am not yet convinced, but is of course something easy to do (just that one has to remember to update the year every year).

To be sure we talk about the same thing, this is unrelated to this PR and that line would only talk about Manopt and its developer, not Ipopt

Reply:
> To be sure we talk about the same thing, this is unrelated to this PR and that line would only talk about Manopt and its developer, not Ipopt

That‘s clear, that the other files would just cover Manopt.jl – I am mainly surprised, that the current file requires a copyright as if it was just copied over here?

But sure if we have the Ipopt line in this file we should also add the Manopt line I think, mentioning the two of us (me as main dev here, you as the clever guy being able to combine both)?

Reply:
> which nearly reads (or for me reads) as plain copyright.

I think you're confusing two subtly different things. To clarify:

 * Copyright protection applies irrespective of whether you write `Copyright (c)`
 * A file without `Copyright (c)` and without a license cannot be redistributed
 * The MIT license grants permission to use and redistribute the file, _provided the copyright notice and MIT license is retained_

From the text of the Ipopt license:

```
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
```

Now we don't need to include the permission notice because Manopt is also MIT. But we do need to include the copyright notice identifying the original authors.

> About having such a line in every code file, hm, I am not yet convinced, but is of course something easy to do (just that one has to remember to update the year every year).

This is another subtle issue. Because copyright applies automatically, and irrespective of the "Copyright (c) YEAR" notice, the year is irrelevant. In JuMP-dev we've just standardized on the same header for every file with the first year that the code was written. You don't need to update every year.

I guess to summarize:

 * Since copyright applies automatically, the only thing that really matters is the license. In this case, both are MIT so you're fine.
 * But including the copyright header in each file means that when someone copies the file to a new repository, they're more likely to leave it in, and it's harder for them to claim that "they didn't realize" they also need to include the line idenfying copyright of the original authors.

Reply:
> I think you're confusing two subtly different things

No, I just missed that it was MIT license, because that line did not mention that and you have to go back to the repository (and yes I did not do that) to check.

Your standardised way seems reasonable, maybe I start with that idea as well. Until now – I never saw much use in having a single file of this repo somewhere else.

So maybe then my only remark here left is: I am still surprised that when Benoit writes so much of the code – it is still a note from someone else that pertains (but maybe the initial template is super important). And Yes I would like to have the second line we discussed earlier then here as swell.